src/gui.cpp

```diff
@@ -0,0 +1,1146 @@
+        }
+        if (set_hot) {
+            ctx->hot_id = gui_id(label);
+            ctx->hot_win_ix = gui_turtle(ctx)->window_ix;
+        }
+    }
+}
+
+bool gui_is_hot(GuiContext *ctx, const char *label)
+{
+    return ctx->last_hot_id == gui_id(label);
+}
+
+void gui_set_active(GuiContext *ctx, const char *label)
+{
+    ctx->active_id = gui_id(label);
+    ctx->hot_id = 0; // Prevent the case where hot becomes assigned some different (overlapping) element than active
+}
+
+void gui_set_inactive(GuiContext *ctx, const char *label)
+{
+    if (ctx->active_id == gui_id(label))
+        ctx->active_id = 0;
+}
+
+bool gui_is_active(GuiContext *ctx, const char *label)
+{
+    return ctx->active_id == gui_id(label);
+}
+
+void gui_button_logic(GuiContext *ctx, const char *label, V2i pos, V2i size, bool *went_up, bool *went_down, bool *down, bool *hover)
+{
+    if (went_up) *went_up = false;
+    if (went_down) *went_down = false;
+    if (down) *down = false;
+    if (hover) *hover = false;
+
+    bool was_released = false;
+    if (gui_is_active(ctx, label)) {
+        if (ctx->key_state[GUI_KEY_LMB] & GUI_KEYSTATE_DOWN_BIT) {
+            if (down) *down = true;
+        }
+        else if (ctx->key_state[GUI_KEY_LMB] & GUI_KEYSTATE_RELEASED_BIT) {
+            if (went_up) *went_up = true;
+            gui_set_inactive(ctx, label);
+            was_released = true;
+        }
+    }
+    else if (gui_is_hot(ctx, label)) {
+        if (ctx->key_state[GUI_KEY_LMB] & GUI_KEYSTATE_PRESSED_BIT) {
+            if (down) *down = true;
+            if (went_down) *went_down = true;
+            gui_set_active(ctx, label);
+        }
+    }
+
+    const V2i c_p = px_to_pt(ctx->cursor_pos, ctx->dpi_scale);
+    if (c_p.x >= pos.x &&
+        c_p.y >= pos.y &&
+        c_p.x < pos.x + size.x &&
+        c_p.y < pos.y + size.y) {
+        gui_set_hot(ctx, label);
+        if (hover && (gui_is_hot(ctx, label) || gui_is_active(ctx, label) || was_released)) *hover = true;
+    }
+}
+
+//void do_skinning(GuiContext *ctx, const char *label, V2i pos, V2i size, Color col);
+
+void gui_begin(GuiContext *ctx, const char *label, bool detached)
+{
+    assert(ctx->turtle_ix < MAX_GUI_STACK_SIZE);
+    if (ctx->turtle_ix >= MAX_GUI_STACK_SIZE)
+        ctx->turtle_ix = 0; // Failsafe
+
+    GuiContext_Turtle *prev = &ctx->turtles[ctx->turtle_ix];
+    GuiContext_Turtle *cur = &ctx->turtles[ctx->turtle_ix + 1];
+
+    ++ctx->turtle_ix;
+    GuiContext_Turtle new_turtle = {};
+    new_turtle.pos = detached ? v2i(0, 0) : gui_pos(ctx, label, prev->pos);
+    new_turtle.start_pos = new_turtle.pos;
+    new_turtle.bounding_max = new_turtle.pos;
+    new_turtle.last_bounding_max = new_turtle.pos;
+    new_turtle.window_ix = prev->window_ix;
+    new_turtle.layer = prev->layer + 1;
+    new_turtle.detached = detached;
+    new_turtle.inactive_dragdropdata = prev->inactive_dragdropdata;
+    snprintf(new_turtle.label, MAX_GUI_LABEL_SIZE, "%s", label);
```

Comment:
currently fails to build on linux/appveyor as its not found, not sure why but because its just %s you can probably just copy label to new_turtle.lable directly?